### PR TITLE
Codebase refactor

### PR DESCRIPTION
**Pull request by cj-zhukov**
## Made refactor of code base:
- spitted code into modules
- load_balancer and worker_host are in separate modules
- worker_host is now custom struct which has method to parse addresses
- utils module has constants and helpers
- using addresses for load balancer and worker hosts by environment variable
- error is a separate module which uses custom enum LoadBalanseError with thiserror and color_eyre
- added some unit tests for worker_hosts

## What's next
- add docker support
- add integration tests
- update hyper to 1.5 version

## Ideas:
- when using decision algos it can be useful to try datafusion with sql query if amount of worker addresses is too big - I'm pretty sure that sql does this job great

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -6,4 +6,9 @@ edition = "2021"
 [dependencies]
 hyper = { version = "0.14.25", features = ["full"] }
 tokio = { version = "1.26.0", features = ["full"] }
-tower = { version = "0.4", features = ["full"] }
\ No newline at end of file
+tower = { version = "0.4", features = ["full"] }
```
Comment (luiscarlosjayk) [inline Cargo.toml]:
```suggestion
tower = { version = "0.5.1", features = ["full"] }
```

File: src/load_balancer/load_balancer.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    pub fn forward_request(&mut self, req: Request<Body>) -> Result<ResponseFuture, LoadBalancerError> {
+        let mut worker_uri = self.get_worker()?.to_owned();
+
+        // Extract the path and query from the original request
+        if let Some(path_and_query) = req.uri().path_and_query() {
+            worker_uri.push_str(path_and_query.as_str());
+        }
+
+        // Create a new URI from the worker URI
+        let new_uri = Uri::from_str(&worker_uri).map_err(|e| LoadBalancerError::InvalidUri(e))?;
```
Comment (luiscarlosjayk) [inline src/load_balancer/load_balancer.rs]:
```suggestion
        let new_uri = Uri::from_str(&worker_uri).map_err(LoadBalancerError::InvalidUri)?;
```

File: src/load_balancer/load_balancer.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        // Extract the headers from the original request
+        let headers = req.headers().clone();
+
+        // Clone the original request's headers and method
+        let mut new_req = Request::builder()
+            .method(req.method())
+            .uri(new_uri)
+            .body(req.into_body())
+            .map_err(|e| LoadBalancerError::HttpError(e))?;
```
Comment (luiscarlosjayk) [inline src/load_balancer/load_balancer.rs]:
```suggestion
            .map_err(LoadBalancerError::HttpError)?;
```

File: src/load_balancer/worker_hosts.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            if Self::validate(&address)? {
+                hosts.push(address);
+            }
+        }
+
+        Ok(Self { hosts })
+    }
+
+    fn validate(address: &str) -> Result<bool, LoadBalancerError> {
+        let re = Regex::new(r"^http://([a-zA-Z0-9.-]+):([0-9]{1,5})$")?;
```
Comment (thedrummeraki) [inline src/load_balancer/worker_hosts.rs:26]: This is pretty cool. Something I was planning on doing in my implementation was using type-based validation, similarly to how we'd have an `Email` struct that validates the contents on parse 👌 So you would catch earlier errors with the struct + not have to `validate` anywhere else in the code (even though it's just once right now 😅)... Just a thought, and could be a good thing to do in an upcoming PR.

**Pull request merged**